### PR TITLE
Elasticsearch target

### PR DESCRIPTION
**Pull request by ademar**
working prototype

### CODE REVIEW

**Comment by haf** [comment]
Nice, looks good. Let me have a deeper look tonight. Are you planning on doing more on it or do you feel it's ready?

Also, have you considered how you're going to handle ES's typed indicies and index rotation? Index per day?

**Comment by ademar** [comment]
It is complete enough for my current usage. There aren't really many options here.

At the moment I have hard-coded the index as "logary-yyy.mm.dd" (daily rotation)  and the `type` can be overridden in the `create` call, by default `type` is "logs". I don't see a way of generalizing the index without writing a lot of code. Suggestions are welcome.

File: src/targets/Logary.Targets.ElasticSearch/Targets_ElasticSearch.fs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    let request = 
+      Request.createUrl Post endpointUrl
+      |> Request.body (RequestBody.BodyRaw bytes)
+
+    Request.responseAsString request
+
+  let loop (conf : ElasticSearchConf)
+           (ri : RuntimeInfo)
+           (requests : RingBuffer<_>)
+           (shutdown : Ch<_>) =
```
Comment (haf) [inline src/targets/Logary.Targets.ElasticSearch/Targets_ElasticSearch.fs]: I think you need to handle shutdown too, then I think we can merge.

File: src/targets/Logary.Targets.ElasticSearch/Targets_ElasticSearch.fs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+open Logary.Utils.Chiron
+
+/// This is the default address this Target publishes messages to.
+[<Literal>]
+let DefaultPublishTo =
+  "http://localhost:9200"
+
+type ElasticSearchConf =
+  { publishTo  : string
+    _type : string }
```
Comment (haf) [inline src/targets/Logary.Targets.ElasticSearch/Targets_ElasticSearch.fs]: Some `///` docs?

File: src/targets/Logary.Targets.ElasticSearch/Targets_ElasticSearch.fs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    init conf
+
+let create conf = TargetUtils.stdNamedTarget (Impl.loop conf)
+
+/// Use with LogaryFactory.New( s => s.Target<ElasticSearch.Builder>() )
+type Builder(conf, callParent : FactoryApi.ParentCallback<Builder>) =
+
+  /// Specifies the ElasticSearch url.
+  member x.PublishTo(publishTo : string) =
```
Comment (haf) [inline src/targets/Logary.Targets.ElasticSearch/Targets_ElasticSearch.fs]: Could expose `_type` in the fluent API too

File: src/targets/Logary.Targets.ElasticSearch/Targets_ElasticSearch.fs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      |> Request.body (RequestBody.BodyRaw bytes)
+
+    Request.responseAsString request
+
+  let loop (conf : ElasticSearchConf)
+           (ri : RuntimeInfo)
+           (requests : RingBuffer<_>)
+           (shutdown : Ch<_>) =
+
+    let rec init config =
```
Comment (haf) [inline src/targets/Logary.Targets.ElasticSearch/Targets_ElasticSearch.fs]: Could be made a single function.

File: src/targets/Logary.Targets.ElasticSearch/Targets_ElasticSearch.fs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+module internal Impl =
+
+  open HttpFs.Client
+
+  let generateId =
+    let chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWUXYZ0123456789"
+    let charsLen = chars.Length
+    let random = System.Random()
+    fun len -> 
+        let randomChars = [|for i in 0..len -> chars.[random.Next(charsLen)]|]
```
Comment (haf) [inline src/targets/Logary.Targets.ElasticSearch/Targets_ElasticSearch.fs]: 2 space indentation

**Comment by haf** [comment]
Added some comments. Think it looks good.

With index I meant that ES interprets the values as things like 'date' or 'numeric' and then if you send something that doesn't match (think; initial message is the string "3" then next message "hi", you might have a problem).

**Comment by haf** [comment]
Sweet! Thanks for your target, it's much appreciated Ademar :)

**Pull request merged**